internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go

```diff
@@ -0,0 +1,321 @@
+	reconciler workers.Reconciler,
+	dataplaneClusterConfig *config.DataplaneClusterConfig,
+	clusterProvidersConfig *config.ProviderConfig,
+	kafkaConfig *config.KafkaConfig,
+	clusterService services.ClusterService,
+) *DynamicScaleDownManager {
+
+	return &DynamicScaleDownManager{
+		BaseWorker: workers.BaseWorker{
+			Id:         uuid.New().String(),
+			WorkerType: dynamicScaleDownWorkerType,
+			Reconciler: reconciler,
+		},
+
+		dataplaneClusterConfig: dataplaneClusterConfig,
+		clusterProvidersConfig: clusterProvidersConfig,
+		kafkaConfig:            kafkaConfig,
+		clusterService:         clusterService,
+	}
+}
+
+func (m *DynamicScaleDownManager) Start() {
+	m.StartWorker(m)
+}
+
+func (m *DynamicScaleDownManager) Stop() {
+	m.StopWorker(m)
+}
+
+func (m *DynamicScaleDownManager) Reconcile() []error {
+	var errList fleeterrors.ErrorList
+	if !m.dataplaneClusterConfig.IsDataPlaneAutoScalingEnabled() {
+		glog.Infoln("dynamic scaling is disabled. Dynamic scale down reconcile event skipped")
+		return nil
+	}
+	glog.Infoln("running dynamic scale down reconcile event")
+	defer m.logReconcileEventEnd()
+
+	err := m.processDynamicScaleDownReconcileEvent()
+	if err != nil {
+		errList.AddErrors(err)
+	}
+
+	return errList.ToErrorSlice()
+}
+
+func (m *DynamicScaleDownManager) processDynamicScaleDownReconcileEvent() error {
+	var errList fleeterrors.ErrorList
+	kafkaStreamingUnitCountPerClusterList, err := m.clusterService.FindStreamingUnitCountByClusterAndInstanceType()
+
+	if err != nil {
+		errList.AddErrors(err)
+		return errList
+	}
+
+	type processed struct {
+		indexesOfStreamingUnitForSameClusterID []int
+		processed                              bool
+	}
+
+	var alreadyProcessedClusters map[string]processed = map[string]processed{}
+
+	for i, kafkaStreamingUnitCountPerCluster := range kafkaStreamingUnitCountPerClusterList {
+		clusterID := kafkaStreamingUnitCountPerCluster.ClusterId
+		existing, ok := alreadyProcessedClusters[clusterID]
+
+		if !ok {
+			alreadyProcessedClusters[clusterID] = processed{
+				indexesOfStreamingUnitForSameClusterID: []int{i},
+				processed:                              false,
+			}
+		} else {
+			alreadyProcessedClusters[clusterID] = processed{
+				indexesOfStreamingUnitForSameClusterID: append(existing.indexesOfStreamingUnitForSameClusterID, i),
+				processed:                              existing.processed,
+			}
+		}
+	}
+
+	for _, suCount := range kafkaStreamingUnitCountPerClusterList {
+		clusterID := suCount.ClusterId
+		existing, alreadyProcessed := alreadyProcessedClusters[clusterID]
+		if alreadyProcessed && existing.processed { // skip already processed
+			continue
+		}
+
+		alreadyProcessedClusters[clusterID] = processed{
+			indexesOfStreamingUnitForSameClusterID: existing.indexesOfStreamingUnitForSameClusterID,
+			processed:                              true,
+		}
+
+		if suCount.Status != api.ClusterReady.String() { // do not consider non ready cluster for scale down evaluation
```

Comment:
Maybe we can apply cluster filtering above on kafkaStreamingUnitCountPerClusterList ?

Reply:
At this line https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/4627eb783d268e54ca1da2102f22b33ac367f9e5/internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go#L100 
with something like
```
processed: kafkaStreamingUnitCountPerCluster.Status != api.ClusterReady.String()
```
?

Reply:
No. I mean just remove the entries from the original list

Reply:
Maybe it's not feasible if we depend on having complete information

Reply:
I ended up considering them as already processed. 

We could filter them, as well since over here https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/6c071fb6699f56cbc3dc379abdaf1c9a9fa15df8/internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go#L246 we are already filtering some of them (the status in terraforming states). That could be a potential future perf improvement

Reply:
I am not sure that's possible as dynamic scale up mgr needs the global information about all clusters for a given cloudprovider-cloudproviderregion-instancetype

Reply:
discussed on another comment.